multidim-interop/rust/v0.50/Makefile

```diff
@@ -0,0 +1,18 @@
+image_name := rust-v0.50
+commitSha = 53e477d4423106bb9f9e744a4b275dfb6213ca0a
```

Comment:
@thomaseizinger / @jxs

This is how I imagine defining future released versions. You can use a published version instead. I wasn't able to figure out how to get that working locally for me.

Also a note, from testing this manually. The Rust listener I don't think is listening for a sigint. The listener doesn't exit when compose tries to stop it after the dialer finishes.

Reply:
also this isn't for the released version 0.50, it's for master I believe. That needs to be fixed, but the test-to-spec update isn't backported yet. Easy to do, I just haven't done it yet.

Reply:
So if I understand correctly, we download the archive and build the container from the source of the tag?

Or do you expect the test to be defined here and we just reference the downloaded source-code?

Reply:
If the test is still co-located with the source-code, I don't really see the benefit of doing this. Can you lay it out please?

If the test is not co-located but defined in here, then we can just use the package manager to point the released version.

Reply:
The test is colocated with the implementation (in rust-libp2p). 

The code here should be minimal, fetch the repo at the specified hash/tag and then build the container. 

The dockerfile here is simply because I couldn't use the one in the rust-libp2p repo since im on a Mac and haven't setup cross builds. 

This way you can run these tests locally without having access to the registry. 

CI will cache this as well so it shouldn't add much time relative to a registry either.

Reply:
> This way you can run these tests locally without having access to the registry.

Is that the only requirement that makes the two solutions differ? Because you can also do that if you just build the container locally and give it the right tag, then docker won't try to pull it again.

Reply:
Interesting. What's the right tag in this case?

Reply:
> CI will cache this as well so it shouldn't add much time relative to a registry either.

At least with Rust, caching and docker are a tricky business. Need to create a separate layer for the dependencies and any change to the lockfile will invalidate that one again, adding a solid 5min or so to the build. The cache we have on CI on the other hand can handle lockfile changes and only rebuilds what is necessary. Hence the solution of copying the binary in.

Reply:
> Interesting. What's the right tag in this case?

This one: https://github.com/libp2p/rust-libp2p/blob/53e477d4423106bb9f9e744a4b275dfb6213ca0a/.github/workflows/publish-interop-test-images.yml#L48

Expanded example: `ghcr.io/libp2p/rust-libp2p/ping-interop-test:v0.50.0`

Try the following locally:

- `docker run hello-world`
- Disconnect from the internet
- `docker run hello-world`

The first will pull the container but unless you specify `--pull=always` on the second one, it will default to `missing` and because you have it, run your local container.

Reply:
Following from that, once you've run the test-suite once locally, it will work without a network connection.

Reply:
```
❯ docker run --rm -it ghcr.io/libp2p/rust-libp2p/ping-interop-test:v0.50.0                                                                                            
Unable to find image 'ghcr.io/libp2p/rust-libp2p/ping-interop-test:v0.50.0' locally
docker: Error response from daemon: Head "https://ghcr.io/v2/libp2p/rust-libp2p/ping-interop-test/manifests/v0.50.0": denied.
```

I assume I need to authenticate with the registry somehow?

There are two problems I have with just relying on the tag.
1. I can't checkout this repo and run the test. Now I need to setup a registry to get this working.
2. It makes changes to the test implementation harder. Do I have to have a GH action publish a new container for that change? I can't do what the action does [here](https://github.com/libp2p/rust-libp2p/blob/53e477d4423106bb9f9e744a4b275dfb6213ca0a/.github/workflows/publish-interop-test-images.yml#L48) because I would need to cross compile to linux to create the binary.

My requirements:
1. It should be reproducible from a given test-plans commit. If a commit passes, it should always pass in the future.
2. I should be able to run the tests locally, and be able to tweak tests locally.

I'm happy to cache things all day long. And if we could use the registry as a cache, I'm fine with doing that as well. That's what I thought you meant when you said "Because you can also do that if you just build the container locally and give it the right tag". But now I understand you meant that I could manually build it and tag it myself so it doesn't try to fetch from the registry. This breaks the reproducibility.

Reply:
> ```
> ❯ docker run --rm -it ghcr.io/libp2p/rust-libp2p/ping-interop-test:v0.50.0                                                                                            
> Unable to find image 'ghcr.io/libp2p/rust-libp2p/ping-interop-test:v0.50.0' locally
> docker: Error response from daemon: Head "https://ghcr.io/v2/libp2p/rust-libp2p/ping-interop-test/manifests/v0.50.0": denied.
> ```
> 
> I assume I need to authenticate with the registry somehow?

Nothing is published yet as I wanted to await the final spec of the tests! :)
The branches are ready in rust-libp2p. Just need to click the publish button in the workflow.
 
> There are two problems I have with just relying on the tag.
> 1. I can't checkout this repo and run the test. Now I need to setup a registry to get this working.

Nothing needs to be set up once the images are published. The images will be publicly accessible.

> 2. It makes changes to the test implementation harder. Do I have to have a GH action publish a new container for that change? I can't do what the action does [here](https://github.com/libp2p/rust-libp2p/blob/53e477d4423106bb9f9e744a4b275dfb6213ca0a/.github/workflows/publish-interop-test-images.yml#L48) because I would need to cross compile to linux to create the binary.

The fact that we build the binary outside the container is a simplification (and performance optimisation) that us `rust-libp2p` maintainers accepted because we all run linux. If necessary, we can switch to building inside a container. It has at the end of the day nothing to do with pulling images from a registry or building them locally.

> My requirements:
> 1. It should be reproducible from a given test-plans commit. If a commit passes, it should always pass in the future.

We can (and should) pin the container hash in the `versions.ts` file. That will fetch the right container even if the tag gets overwritten. IMO this is a super strong commitment to reproducibility because we are running bit-for-bit the same binary on every test.

> 2. I should be able to run the tests locally, and be able to tweak tests locally.

Changing the test is tricky because it lives in another source-code repository anyway. For local dev, I'd suggest to build the container locally and add it via `--extra-version`. I don't think we should "cheat" the local setup by overwriting an existing test or container.

Reply:
Let's chat sync. I think that'll be faster than all this back and forth :)

Reply:
@MarcoPolo and I agreed in a sync-chat that using these caches is basically equivalent to a registry, just slightly leaner.

Building a docker container will reuse local / cached layers if they are present. Running an image will pull it unless it is present. If we have a central cache in S3, building and pulling from a registry are essentially equivalent, we are just downloading layers.

The benefit of building and caching over pulling is that we only need to define how to build a container in the implementation repository. Then, any commit (or tag for that matter) can be referenced within the test-plans repo here and used to run the tests.